File: 202.快乐数.js
/*
 * @lc app=leetcode.cn id=202 lang=javascript
 *
 * [202] 快乐数
 */

// @lc code=start
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
  let sum,
    map = new Map();
  while (sum !== 1) {
    sum = 0;
    while (n) {
      sum += Math.pow(n % 10, 2);
      n = Math.floor(n / 10);
    }
    if (map.has(sum)) return false;
    n = sum;
    map.set(sum, 1);
  }
  return true;
};
// @lc code=end

File: store/202.快乐数.c
/*
 * @lc app=leetcode.cn id=202 lang=c
 *
 * [202] 快乐数
 */

// @lc code=start

bool isHappy(int n)
{
  int *arr, m, i, sum;
  arr = (int *)malloc(sizeof(int) * 1000);
  m = n;

  for (i = 0; i < 1000; i++)
    arr[i] = 0;

  while (m != 1)
  {
    sum = 0;
    while (m / 10 != 0 || (m % 10 != 0 && m / 10 == 0))
    {
      sum += (m % 10) * (m % 10);
      m = m / 10;
    }
    arr[sum] += 1;
    for (i = 0; i < 1000; i++)
      if (arr[i] > 1)
        return false;
    m = sum;
  }

  return true;
}
// @lc code=end
